pkg/images/integration/set_impl.go

```diff
@@ -56,18 +56,33 @@ func (e *setImpl) UpdateImageIntegration(integration *storage.ImageIntegration)
 		return
 	}
 
+	var isRegistry bool
+	var isScanner bool
 	for _, category := range integration.GetCategories() {
 		switch category {
 		case storage.ImageIntegrationCategory_REGISTRY:
+			isRegistry = true
 			err = e.registrySet.UpdateImageIntegration(integration)
 		case storage.ImageIntegrationCategory_SCANNER:
+			isScanner = true
 			err = e.scannerSet.UpdateImageIntegration(integration)
 		case storage.ImageIntegrationCategory_NODE_SCANNER: // This is because node scanners are implemented into image integrations
 		default:
 			err = fmt.Errorf("source category %q has not been implemented", category)
 		}
 	}
 
+	// An integration may have a category removed, for example, if an integration went from being
+	// both a registry + scanner to just a registry. On update we need to remove the integration
+	// from the sets it should no longer be a part of.
+	if !isRegistry {
+		_ = e.registrySet.RemoveImageIntegration(integration.GetId())
+	}
+
+	if !isScanner {
+		_ = e.scannerSet.RemoveImageIntegration(integration.GetId())
+	}
+
```

Comment:
If the update failed the remove wouldn't happen, because the `bool` would be `true`.

For example: `integration` has `REGISTRY` category, `registry.Set` update fails, `isRegistry` will be true, therefore integration will NOT be removed from registry set.

Reply:
Related - this function has 'existing' behavior to eclipse errors if multiple categories exists on an integration, only the last one encountered is returned. Could benefit from being an error list instead.

Reply:
https://github.com/stackrox/stackrox/pull/8939/commits/35f0ce0f139373869c4ef43f5205f019944a8b99

Reply:
> Related - this function has 'existing' behavior to eclipse errors if multiple update errors occur, only the last one encountered is returned. I'll switch this to an error list.

That's a nice addition.

Reply:
I see what you mean. Current state / prior to this change a partial update is possible (ie: registrySet updated when scannerSet fails), cross set updates are not thread safe, and the set usages are largely disjointed. 

Based on that I don't see a problem with that partial update, could argue this change is making the update 'more precise'.

Reply:
> The partial update scenario caused by the removal of integrations from their sets was not possible before.

True.  What I meant is that currently the sets will be in an invalid state 100% of the time when removing a category, now they will only be in an invalid state if a partial failure scenario occurs / edge case.

Reply:
if it helps, at the moment, `err = nil` always (just `delete(map, id)`), so we shouldn't end up with any partial errors